Consolidate handling of emulated langinfo items into a single low-level function

khwilliamson opened the pull request:
When nl_langinfo() is missing from a platform, as notably it isn't present in Windows, perl tries to emulate it.

I have long been unhappy with the situation where the default values get returned at a higher level function, and those that get calculated are done in a lower level one.  The default values get returned for some Configurations where there is no variance in results, or where missing platform abilities preclude our calculating things.

This series of mostly bite-size commits rectifies this, by moving everything into a low-level function.  This has a couple of maintainability benefits

- All the code for an item is in the same place, leading to better understandability 
- The low level function has always been the one that gets used internally.  Prior to this, it couldn't be called without an `#ifdef` to make sure it was usable.  Most such calls needed the same `#ifdef` anyway, but it still had to be taken into consideration.  After these commits, that function can be called without needing to think about it.
- There was some duplication, now eliminated

The commits related to `LC_TIME` have been split into tiny related chunks, as one big one would lead to not being able to understand what's going on so easily

khwilliamson committed:
locale.c: Recompute variable in 2nd part of function

This is in preparation for the next commit splitting the function in
two, and the current 2nd part will not need or have this variable passed
in as an argument.

khwilliamson committed:
locale.c: Split a function in two

Prior to this commit, the function was #ifdef'd so that one or the other
part would compile.  But future commits will cause them both to need to
be compiled in some circumstances, with the first part potentially
calling the second part.

Note that the 2nd part doesn't need the parameter 'cat_index' that the
first part does.

khwilliamson committed:
locale.c: Refactor two nested switch() statements

This slightly refactors these so that many case labels are no longer
repeated.  It does this by making the nested one the default: of the
outer one, and the outer's previous default: becomes the inner's
default (which formerly was useless, so panicked).

khwilliamson committed:
locale.c: Move GCC DIAG IGNORE

The previous commit changed two nested switch() statements.  The inner
one turned off implicit switch() case statement fallthrough warnings.
This commit applies that to the entire outer switch().

khwilliamson committed:
locale.c: Handle some ignored categories in langinfo emulation

It is possible to Configure perl to not allow any or all of the locale
categories on the platform to be changed away from the C locale,
generally due to libc limitations.  OpenBSD, for example, keeps every
category but one in C; so perl shouldn't even try to change any of
those.

On platforms without the libc nl_langinfo() function, perl emulates it.
Prior to this commit, the emulation did not properly handle the case
of categories needing to stay in C.  This did not lead to bugs, because
external to this file, any such calls were intercepted by
Perl_langinfo8(), and internally, we just didn't call it under those
circumstances.

But this leads to some awkwardness, and it is more maintainable to have
the handling in one place; which is this low-level emulation.

This commit causes the emulation to handle LC_CTYPE, LC_NUMERIC, and
LC_MONETARY needing to stay in the C locale.  Future commits will add
the other categories, and remove the redundant checks in
Perl_langinfo8().

khwilliamson committed:
locale.c: Tighten circumstances some code is generated

This code need not be compiled except under the circumstances defined in
this commit.

khwilliamson committed:
locale.c: Move check for unknown langinfo items

This moves this check into the lowest level function, so that no
duplication is needed.

khwilliamson committed:
locale.c: Consolidate handling of some langinfo items

Without nl_langinfo(), some Configurations handled nl_langinfo() items
associated with the LC_CTYPE, LC_MONETARY, and LC_NUMERIC categories in
Perl_langinfo8(), and some Configurations in emulate_langinfo().  This
commit consolidates all the handling to emulate_langinfo().

This commit removes the only use of a macro, which is hence also
removed.

khwilliamson committed:
locale.c: Move handling of LC_MESSAGES langinfo items

This commit moves the handling of langinfo items associated with the
LC_MESSAGES category to emulate_langinfo() when nl_langinfo() is not
available on the platform away from Perl_langinfo8().

This is a step in the process of handling all such items in
emulate_langinfo().

khwilliamson committed:
locale.c: Consolidate handling of DAY langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Consolidate handling of ABMON langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Consolidate handling of MON langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Consolidate handling of AM/PM langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Move handling of ERA langinfo item

This commit moves the handling of the ERA langinfo item to
emulate_langinfo() away from Perl_langinfo8().

khwilliamson committed:
locale.c: Consolidate handling of ERA format langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Consolidate handling of ALT_DIGITS langinfo item

Without nl_langinfo(), some Configurations handled this nl_langinfo()
item associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: Remove redundant handling of the AM_PM format

This format is identical whether or not LC_TIME is usable on this
platform or not.  Previously, it was actually handled in two functions,
emulate_langinfo() when LC_TIME is usable, and Perl_langinfo8() when
not.  Previous commits have enabled the format handling in
emulate_langinfo() to be returned instead of having separate cases,
depending on LC_TIME's availability.  That means the handling of this
format in Perl_langinfo8() is redundant and this commit removes it.

khwilliamson committed:
locale.c: Consolidate handling of remaining TIME items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

khwilliamson committed:
locale.c: No strftime() implies empty formats

When returning an nl_langinfo() format when we are emulating
nl_langinfo() when there is no strftime() on the system, return an
empty format.  A non-empty return is useless.

khwilliamson committed:
locale.c: Extract code into a separate function

This new function does not have to be compiled in all Configurations,
and placing it separately removes some #ifdefs, and allows its absence
to be more transparent from its callers.

khwilliamson committed:
locale.c: Can now call my_langinfo_i() in all circumstances

Prior to this commit, the caller had to know if the category in question
was being ignored in this Configuration.  Now, that is transparent.

khwilliamson committed:
locale.c: Always compile save__to_buffer()

The next commit will want to use it in more Configurations, and a future
commit in all Configurations.

khwilliamson committed:
locale.c: langinfo: Better handle unknown input item

On systems with <langinfo.h>, the 'item' parameter here must be a member
of an enum.  If it isn't something we recognize, it's a real problem.

On other systems any int can be passed, and it could just be that the
caller got a wrong value.  For these, we don't know if it is a perl
error or a user error, so just set EINVAL and return failure

khwilliamson committed:
locale.c: Consolidate handling of ABDAY langinfo items

Without nl_langinfo(), some Configurations handled these nl_langinfo()
items associated with the LC_TIME category in Perl_langinfo8(), and some
Configurations in emulate_langinfo().  This commit consolidates all the
handling to emulate_langinfo().

tonycoz commented on embed.fnc:
```diff
@@ -4413,6 +4413,18 @@ S	|void	|ints_to_tm	|NN struct tm *my_tm			\
+				|const line_t immediate_caller_line	\
+				|NN const char *higher_caller_file	\
+				|const line_t higher_caller_line
+S	|void	|set_save_buffer_min_size				\
```
The commits message mentions save_to_buffer(), but this moves only set_save_buffer_min_size() and setlocale_failure_panic_via_i() 

The change to locale.c moves set_save_buffer_min_size() and save_to_buffer() so there's some inconsistency here.

The commit message mentions only save_to_buffer(), I assume you're treating set_save_buffer_min_size() and an implementation detail there, but it doesn't explain setlocale_failure_panic_via_i().

khwilliamson replied:
I think this is a result of `embed.fnc` now being a generated file, which has some downsides.  I moved the two functions `save_to_buffer` and `set_save_buffer_min_size` inside `embed.fnc`, and then the regen phase alphabetically sorts things within each grouping, and the `diff` looks for the minimal diff, which  doesn't necessarily reflect the actual changes from a human perspective.

tonycoz replied:
I don't think that explains it.  If I build with `-Accflags=-DNO_LOCALE` and `make locale.i` then the preprocessed result only includes the definition of save_to_buffer() from locale.c and doesn't include the declaration from proto.h.

I can edit embed.fnc and move save_to_buffer() outside of the `#ifdef USELOCALE` (which appears to be the intent here, similar to set_save_buffer_min_size(), make regen and then locale.i and I see both the save_to_buffer() and set_save_buffer_min_size() declarations from proto.h

Note I inserted save_to_buffer() into it's new position in alphabetical order and the make regen didn't move it.

Note that the setlocale_failure_panic_via_i() declaration is also visible in locale.i when built with `-DNO_LOCALE` as above, which doesn't seem correct.

khwilliamson committed:
locale.c: Compile S_emulate_langinfo() under more Configurations

This is in preparation for some code to be consolidated into just this
one function.

khwilliamson committed:
locale.c: Effectively white-space, comments only

The previous few commits have added/removed code blocks and #if
directives.  Adjust white space accordingly

This also adds explanatary comments

khwilliamson commented:
Then the problem was in `git rebase -i`; either my mistake, or the not uncommon mistake of `rerere`.  I redid that commit and looked at all of them in this sequence.  I found several other things where the rebase had been screwed up; though none as consequential as the one @tonycoz ++ spotted.
